src/statsmodels/contrasts.jl

```diff
@@ -0,0 +1,143 @@
+## Specify contrasts for coding categorical data in model matrix. Contrast types
+## are constructed with data and other options, creating a subtype of
+## AbstractContrast. AbstractContrast provides the interface for creating modle
+## matrix columns and coefficient names
+##
+## ModelFrame will hold a Dict{Symbol, T<:AbstractContrast} that maps column
+## names to contrasts.
+##
+## ModelMatrix will check this dict when evaluating terms, falling back to a
+## default for any categorical data without a specified contrast.
+
+abstract AbstractContrast
+
+termnames(term::Symbol, col::Any, contrast::AbstractContrast) =
+    ["$term - $name" for name in contrast.termnames]
+
+function cols(v::PooledDataVector, contrast::AbstractContrast)
+    ## make sure the levels of the contrast matrix and the categorical data
+    ## are the same by constructing a re-indexing vector. Indexing into
+    ## reindex with v.refs will give the corresponding row number of the
+    ## contrast matrix
+    reindex = [findfirst(l .== contrast.levels) for l in levels(v)]
+
+    ## TODO: add kwarg for full-rank contrasts (e.g., when intercept isn't
+    ## specified in a model frame).
+
+    return contrast.matrix[reindex[v.refs], :]
+end
+
+## Default contrast is treatment coding:
+cols(v::PooledDataVector) = cols(v, TreatmentContrast(v))
+termnames(term::Symbol, col::PooledDataArray) = termnames(term, col, TreatmentContrast(col))
+
+
+## Constructing a contrast from a non-pooled data vector will first pool it
+## 
+## (NOT SURE THIS IS GOOD: constructing columns also depends on levels, so for
+## _that_ to work would need to somehow hold onto pooled data...)
```

Comment:
Honestly, I'm not sure what I was thinking now when I wrote that :)  

There was some discussion in #867 about a mechanism for treating continuous variables as categorical. One way to do this would be to provide a contrast for that column, and a contrast constructor that takes a column that's _not_ a `PooledDataArray`, calls `pool` on the column before evaluating the contrast (like here). I think I was worried that somehow the levels of the new pooled vector wouldn't be held onto, but that's wrong (they're stored in the contrast object that's created when the `ModelFrame` constructor calls the contrast constructors).

The thing that's missing here is a `cols` method that takes a non-`PooledDataArray` column and a contrast and handles that. The problem there is that you're calling `pool` twice (at least). The alternative is somehow to store the pooled version of the column but that just seems really messy to me.